src/mappings/utils/pools.ts

```diff
@@ -152,17 +186,200 @@ function compareCurrencies(currency0: Currency, currency1: Currency): number {
     
-    const order: number = compareCurrencies(currency0, currency1);
-    // swap strings around if needed
-    if (order > 0) {
-        // works because strings (ie. not the capitalized String) are passed by value
-        const hold = firstCurrencyString;
-        firstCurrencyString = secondCurrencyString;
-        secondCurrencyString = hold;
+    return `(${firstCurrencyString},${secondCurrencyString})`;
+}
+
+function getDexGeneralPairStatusesStorage(ctx: Ctx, block: SubstrateBlock): DexGeneralPairStatusesStorage {
+    const dexGeneralStorage = new DexGeneralPairStatusesStorage(ctx, block);
+    if (!dexGeneralStorage.isExists) {
+        throw Error("DexGeneral.PairStatuses storage is not defined for this spec version");
     }
+    return dexGeneralStorage;
+}
 
-    return `(${firstCurrencyString},${secondCurrencyString})`;
+function getDexStablePoolsStorage(ctx: Ctx, block: SubstrateBlock): DexStablePoolsStorage {
+    const dexStablePoolStorage = new DexStablePoolsStorage(ctx, block);
+    if (!dexStablePoolStorage.isExists) {
+        throw Error("DexStable.Pools storage is not defined for this spec version");
+    }
+    return dexStablePoolStorage;
+}
+
+function getAsBasePool(pool: Pool): BasePool {
+    if (isBasePool(pool)) {
+        return pool.value;
+    } else if (isMetaPool(pool)) {
+        return pool.value.info;
+    } else {
+        // use of any to future-proof for if/when pool types are expanded.
+        throw Error(`getAsBasePool: Unkown pool type [${(pool as any).__kind}], unable to return BasePool data.`);
+    }
+}
+
+export async function buildNewSwapEntity(
+    ctx: Ctx,
+    block: SubstrateBlock,
+    eventId: string,
+    poolType: PoolType,
+    swapDetails: SwapDetails,
+    height: Height,
+    blockTimestamp: Date,
+    poolId?: number
+): Promise<Swap> {
+    let feeRate = Big(0);
+
+    if (poolType == PoolType.Standard) {
+        const currencyCompareValue = compareCurrencies(swapDetails.from.currency, swapDetails.to.currency);
+        const currencyPairKey: [CurrencyId, CurrencyId] = currencyCompareValue < 0 
+            ? [swapDetails.from.currencyId, swapDetails.to.currencyId] 
+            : [swapDetails.to.currencyId, swapDetails.from.currencyId];
+    
+        const dexGeneralStorage = getDexGeneralPairStatusesStorage(ctx, block);
+        
+        if (dexGeneralStorage.isV1021000) {
+            const rawStorage = await dexGeneralStorage.getAsV1021000(currencyPairKey);
+            if (rawStorage.__kind === "Trading") {
+                // raw fee rate is in basis points, so times 0.0001 for actual rate
+                feeRate = Big((rawStorage as PairStatus_Trading).value.feeRate.toString()).mul(0.0001);
+            }
+        } else {
+            ctx.log.warn("UNKOWN STORAGE VERSION: DexGeneral.PairStatuses");
+        }
+    } else if (poolId === undefined) {
+        ctx.log.error("buildNewSwapEntity: no poolId defined, unable to lookup fee rates");
```

Comment:
I would prefer having a swap entity saved even if the fees are incorrect/missing rather than not saving it because we fail to find the pool by id.

Or in other words, I'd rather throw an error than just logging an error and skipping the swap event.
What do you think?

Reply:
> I'd rather throw an error than just logging an error and skipping the swap event. 
That sounds reasonable to me :+1:

Reply:
Done with commit [9222be4](https://github.com/interlay/interbtc-squid/pull/114/commits/9222be4cf88dce31a4ca72ee465fb5e34ec3c6aa)